plugins/tpm/fu-tpm-device.c

```diff
@@ -0,0 +1,232 @@
+{
+	FuTpmDevice *self = FU_TPM_DEVICE (device);
+	TSS2_RC rc;
+	guint32 tpm_type = 0;
+	guint32 version1 = 0;
+	guint32 version2 = 0;
+	g_autofree gchar *family = NULL;
+	g_autofree gchar *id1 = NULL;
+	g_autofree gchar *id2 = NULL;
+	g_autofree gchar *manufacturer = NULL;
+	g_autofree gchar *vendor1 = NULL;
+	g_autofree gchar *vendor2 = NULL;
+	g_autofree gchar *vendor3 = NULL;
+	g_autofree gchar *vendor4 = NULL;
+	g_autofree gchar *vendor_id = NULL;
+	g_autofree gchar *vendor = NULL;
+	g_autofree gchar *version = NULL;
+	g_autoptr(ESYS_CONTEXT) ctx = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	/* setup TSS */
+	rc = Esys_Initialize (&ctx, NULL, NULL);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error_literal (error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND,
+				     "failed to initialize TPM library");
+		return FALSE;
+	}
+	rc = Esys_Startup (ctx, TPM2_SU_CLEAR);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+				     "failed to initialize TPM");
+		return FALSE;
+	}
+
+	/* lookup guaranteed details from TPM */
+	family = fu_plugin_tpm_get_string (ctx, TPM2_PT_FAMILY_INDICATOR, error);
+	if (family == NULL) {
+		g_prefix_error (error, "failed to read TPM family");
+		return FALSE;
+	}
+	manufacturer = fu_plugin_tpm_get_string (ctx, TPM2_PT_MANUFACTURER, error);
+	if (manufacturer == NULL) {
+		g_prefix_error (error, "failed to read TPM manufacturer");
+		return FALSE;
+	}
+	vendor1 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_1, error);
+	if (vendor1 == NULL) {
+		g_prefix_error (error, "failed to read TPM vendor string");
+		return FALSE;
+	}
+	if (!fu_plugin_tpm_get_uint32 (ctx, TPM2_PT_VENDOR_TPM_TYPE, &tpm_type, error)) {
+		g_prefix_error (error, "failed to read TPM type");
+		return FALSE;
+	}
+
+	/* these are not guaranteed by spec and may be NULL */
+	vendor2 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_2, error);
+	vendor3 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_3, error);
+	vendor4 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_4, error);
+	vendor = g_strjoin ("", vendor1, vendor2, vendor3, vendor4, NULL);
+	fu_device_set_vendor (device, vendor);
+
+	/* add GUIDs to daemon */
+	id1 = g_strdup_printf ("TPM\\VEN_%s&DEV_%04X", manufacturer, tpm_type);
+	fu_device_add_instance_id (device, id1);
+	id2 = g_strdup_printf ("TPM\\VEN_%s&DEV_%04X&VER_%s", manufacturer, tpm_type, family);
+	fu_device_add_instance_id (device, id2);
+
+	/* enforce vendors can only ship updates for thier own hardware */
+	vendor_id = g_strdup_printf ("TPM:%s", manufacturer);
+	fu_device_set_vendor_id (device, vendor_id);
+
+	/* get version */
+	if (!fu_plugin_tpm_get_uint32 (ctx, TPM2_PT_FIRMWARE_VERSION_1, &version1, error))
+		return FALSE;
+	if (!fu_plugin_tpm_get_uint32 (ctx, TPM2_PT_FIRMWARE_VERSION_2, &version2, error))
+		return FALSE;
+	version = g_strdup_printf ("%u.%u", version1, version2);
+	fu_device_set_version (device, version, FWUPD_VERSION_FORMAT_PAIR);
```

Comment:
So at least for Nuvoton this isn't a pair - it should be quad.  The spec make allusions that this should be a vendor specific way to interpret this field, but helpfully the spec links to more details are missing.
https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-2-Structures-01.38.pdf

Reply:
> it should be quad

A 64 bit number -> `quad`?

Reply:
Yeah you can see `TPM2_PT_FIRMWARE_VERSION_1` is 0x70002 and `TPM2_PT_FIRMWARE_VERSION_2` is 0x2, which should expand to 7.2.0.2 for the TPM in my system.

Reply:
So we convert each uint32 numbers to pairs, and then glue them together? Maybe you were right about version_raw being 64 bit afterall...

Reply:
Yeah, similarly to how the vendor strings are glued together.  i think it' because all these definitions need to be done in 4 bytes increments.

> Maybe you were right about version_raw being 64 bit afterall...

Hah, I guess maybe so :)  Good thing that's not in a released version yet.

Reply:
Okay, fixed.